### DISCUSSION TITLE
Questions about inference results

### DISCUSSION THREAD

**Discussion by Alex8211101** [Project-MONAI/MONAI#4414] [Q&A]
have tried two methods to generate the predicted results. But there are some problems.
n this section of the program there is a problem that the image does not match the origin of the result.
`test_dicts = [{'image': image_name, 'label': label_name}
                  for image_name, label_name in zip(test_images, test_segs)]

    # define transforms for image and segmentation
    val_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            EnsureChannelFirstd(keys=['image','label']),
            ScaleIntensityd(keys="image"),
            EnsureTyped(keys=["image", "label"]),
        ]
    )
    val_ds = monai.data.Dataset(data=test_dicts, transform=val_transforms)
    # sliding window inference need to input 1 image in every iteration
    val_loader = DataLoader(val_ds, batch_size=1, num_workers=0, collate_fn=list_data_collate)
    dice_metric = DiceMetric(include_background=False, reduction="mean", get_not_nans=False)
    post_trans = Compose([EnsureType(), Activations(softmax=True), AsDiscrete(argmax=True),])
    # post_trans = Compose([EnsureType(), Activations(softmax=True), AsDiscrete(threshold_values=True, logit_thresh=0.5)])
    saver = SaveImage(output_dir="./out/SegResNet/tibia", output_ext=".nii", output_postfix="seg")
    # try to use all the available GPUs
    devices = [torch.device("cuda" if torch.cuda.is_available() else "cpu")]
    model = monai.networks.nets.SegResNet(out_channels=2).cuda()

    model.load_state_dict(torch.load(os.path.join('log/SegResNet/tibia', "best_metric_model.pth")))

    # if we have multiple GPUs, set data parallel to execute sliding window inference
    if len(devices) > 1:
        model = torch.nn.DataParallel(model, device_ids=devices)

    model.eval()
    with torch.no_grad():
        for val_data in val_loader:
            val_data['image'], val_labels = val_data["image"].to(devices[0]), val_data["label"].to(devices[0])
            # define sliding window size and batch size for windows inference
            # val_images = torch.unsqueeze(val_images,dim=0)
            roi_size = (320, 320,128)
            sw_batch_size = 4
            val_outputs = sliding_window_inference(val_data['image'], roi_size, sw_batch_size, model,overlap=0.7)
            val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
            val_labels = decollate_batch(val_labels)
            meta_data = decollate_batch(val_data["image_meta_dict"])
            # compute metric for current iteration
            dice_metric(y_pred=val_outputs, y=val_labels)
            for val_output, data in zip(val_outputs, meta_data):
                saver(val_output, data)
        # aggregate the final mean dice result
        print("evaluation metric:", dice_metric.aggregate().item())
        # reset the status
        dice_metric.reset()`
![H8LV{ZNB0` YX5(K_)}Z8`I](https://user-images.githubusercontent.com/76839078/171200644-74ab71fb-88fe-4266-b508-77e22d490a51.png)

**Comment by Alex8211101** [comment]
In this part of the program, there is a problem that the result is displayed incorrectly.

> test_images = sorted(glob(os.path.join(opt.images_folder, 'test', 'image*.nii')))
test_dicts = [{'image': image_name}for image_name in test_images]
test_org_transforms = Compose(
    [
        LoadImaged(keys="image"),
        EnsureChannelFirstd(keys='image'),
        Orientationd(keys=["image"], axcodes="RAS"),
        EnsureTyped(keys="image"),
    ]
)
test_org_ds = Dataset(
    data=test_dicts, transform=test_org_transforms)
test_org_loader = DataLoader(test_org_ds, batch_size=1, num_workers=0)
post_transforms = Compose([
    EnsureTyped(keys="pred"),
    Activationsd(keys="pred",softmax=True),
    Invertd(
        keys="pred",
        transform=test_org_transforms,
        orig_keys="image",
        meta_keys="pred_meta_dict",
        orig_meta_keys="image_meta_dict",
        meta_key_postfix="meta_dict",
        nearest_interp=False,
        to_tensor=True,
    ),
    AsDiscreted(keys="pred",to_onehot=2,argmax=True,threshold=0.5),
    SplitChanneld(keys="pred", channel_dim=0),
    SaveImaged(keys="pred_0",
               meta_keys="pred_meta_dict",
               output_dir="./out/SegResNet/tibia/",
               output_postfix="seg_0",output_ext='.nii' ,resample=False),
    SaveImaged(keys="pred_1",
               meta_keys="pred_meta_dict",
               output_dir="./out/SegResNet/tibia/",
               output_postfix="label", output_ext='.nii.gz', resample=False),
])
net = monai.networks.nets.SegResNet(out_channels=2)

net.cuda()
net.load_state_dict(torch.load(os.path.join('log/SegResNet/tibia/', "best_metric_model.pth")))
net.eval()
with torch.no_grad():
    for test_data in test_org_loader:
        test_data['image'] = test_data["image"].cuda()
        roi_size = (256,256,96)
        sw_batch_size = 4
        test_data["pred"] = sliding_window_inference(
            test_data['image'], roi_size, sw_batch_size, net)
        test_data = [post_transforms(i) for i in decollate_batch(test_data)]
![image](https://user-images.githubusercontent.com/76839078/171204006-1d1ba7a1-025e-4650-9fec-2ddf118a372e.png)

**Reply by Nic-Ma** [reply]
Hi @Alex8211101 ,

I think maybe you need to add the `invertd` transform for the postprocessing, referring to:
https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/torch/unet_inference_dict.py#L70

Thanks.

**Reply by Alex8211101** [reply]
Thanks for your quick reply,I have used `Invert`,when I tried to use `Orientationd(keys="image", axcodes="RAS"),`in pre_transforms,I got the result like the picture below,is it related to `Orientationd`?
![image](https://user-images.githubusercontent.com/76839078/171318467-8ffcd45e-cc31-4b13-a91d-b01ffadbc080.png)
Thanks.

**Reply by Alex8211101** [reply]
Finally,I found where the problem happened,I used the "cropped‘ labels during training,I retrained and got the correct results,thanks again for your answer.

**Reply by Nic-Ma** [reply]
Cool, glad to see your progress.
